lib/spack/spack/solver/profiler.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Profiling propagator for clingo solves."""
+
+import re
+from typing import Any, Dict, List
+
+import llnl.util.tty as tty
+
+
+class Data:
+    """Counters for propagations and undos of an atom."""
+
+    __slots__ = ["atom", "literal", "prop", "undo"]
+
+    # currently we use Any for clingo types because clingo has a bunch of import
+    # wrappers around it that make typing difficult (see spack.solver.core for details)
+    def __init__(self, atom: Any, literal: int, prop: int, undo: int):
+        self.atom = atom
+        self.literal = literal
+        self.prop = prop
+        self.undo = undo
+
+
+class AggregatedData:
+    """Aggregated data for a profile, constructed from ``Data``.
+
+    We coarsen from atom granularity to string keys when aggregating.
+    """
+
+    __slots__ = ["name", "prop", "undo"]
+
+    def __init__(self, name: str, prop: int, undo: int):
+        self.name = name
+        self.prop = prop
+        self.undo = undo
+
+
+class ProfilePropagator:
+    """Profiling propagator for `spack solve --profile`.
+
+    Register this with the ``clingo.Control`` object to profile a solve.
+    """
+
+    _literal_to_atom: Dict
+    _profile: Dict[int, Data]
+
+    def init(self, init) -> None:
+        self._literal_to_atom = {}
+        self._profile = {}
+        for atom in init.symbolic_atoms:
+            solver_literal = init.solver_literal(atom.literal)
+            self._profile[solver_literal] = Data(atom, solver_literal, 0, 0)
```

Comment:
I'm doing something similar (manually) to try getting some understanding of heuristic. In that case I override the `decide` method.

 I _think_ this is wrong, in the sense that each `solver_literal` may be associated with many `atom`, so the structure should be more:
```suggestion
            self._profile[solver_literal].append(Data(atom, solver_literal, 0, 0))
```

Reply:
Not really understanding this b/c AFAIK each literal maps to exactly one (fully grounded) atom. There's one `Data` per literal/atom. Can you elaborate on this?

If you want, we could track decisions as well -- could it just be a third column?

Reply:
I _think_ the representation of the grounder is not the same as the one of the solver.

Each symbolic atom maps to a solver literal, but they may map to the same one (I guess if they can be proven t be always true/false together). Also, I think there are solver literals that do not map to symbolic atom and are just used as an "internal" representation for the solver.